syntax = "proto3";

package swarm;

import "simulation/swarm/proto/dynamic_config.proto";
import "simulation/swarm/proto/state.proto";
import "simulation/swarm/proto/plotting_config.proto";

// Interceptor type enumeration.
enum InterceptorType {
  MICROMISSILE = 0;
  HYDRA_70 = 1;
}

// Threat type enumeration.
enum ThreatType {
  DRONE = 0;
  MISSILE = 1;
}

// Flight phase enumeration.
enum FlightPhase {
  INITIALIZED = 0;
  READY = 1;
  BOOST = 2;
  MIDCOURSE = 3;
  TERMINAL = 4;
  TERMINATED = 5;
}

// Agent configuration.
message AgentConfig {
  // Submunitions configuration.
  message SubmunitionsConfig {
    // Number of submunitions.
    uint32 num_submunitions = 1;

    // Launch configuration.
    LaunchConfig launch_config = 2;

    // Agent configuration.
    AgentConfig agent_config = 3;
  }

  // Agent type.
  oneof type_oneof {
    // Interceptor type.
    InterceptorType interceptor_type = 1;
    // Threat type.
    ThreatType threat_type = 2;
  }

  // Initial state.
  State initial_state = 3;

  // Initial state standard deviation to randomize the state variables.
  State standard_deviation = 4;

  // Dynamic configuration.
  DynamicConfig dynamic_config = 5;

  // Plotting configuration.
  PlottingConfig plotting_config = 6;

  // Submunitions configuration.
  SubmunitionsConfig submunitions_config = 7;
}

// Agent swarm configuration.
message AgentSwarmConfig {
  // Number of agents.
  uint32 num_agents = 1;

  // Agent configuration.
  AgentConfig agent_config = 2;
}
